phylo/src/tree/nj_builder.rs

```diff
@@ -0,0 +1,750 @@
+    /// let tree = nj_builder.build(&sequences, &mut rng)?;
+    /// assert_eq!(tree.len(), 7);
+    /// assert_eq!(tree.leaves().len(), 4);
+    /// # Ok(()) }
+    /// ```
+    pub fn new_with_softmax(distance_function: D, temperature: f64) -> Self {
+        info!("Creating NJTreeBuilder with the softmax strategy of selecting the best pair of nodes to join to increase stochasticity
+        in tree building. Using temperature {temperature} to interpolate between uniform and softmax distributions.");
+        if temperature > 1.0 {
+            debug!("Temperature should not be greater than 1.0 (set to {temperature}), clamping to 1.0");
+        } else if temperature < 0.0 {
+            debug!(
+                "Temperature should not be less than 0.0 (set to {temperature}), clamping to 0.0"
+            );
+        }
+        let t = temperature.clamp(0.0, 1.0);
+        Self {
+            randomise: Strategy::SoftmaxUniform(t),
+            distance_function,
+        }
+    }
+
+    /// Computes the softmax distribution from the delta tree lengths without temperature
+    /// scaling. As the delta tree lengths are negative, they are inverted before computing
+    /// the distribution, so that smaller (more negative) delta tree lengths have higher probabilities.
+    fn softmax_from_deltas(mut delta_tree_len: DVector<f64>) -> DVector<f64> {
+        // Invert deltas for softmax, most negative delta should have the highest probability
+        delta_tree_len.scale_mut(-1.0);
+        // Avoid copying the matrix by mutating in place
+        for element in delta_tree_len.iter_mut() {
+            *element = element.exp();
+        }
+        delta_tree_len.unscale_mut(delta_tree_len.sum());
+        delta_tree_len
+    }
+
+    /// Computes the softmax probabilities from the delta tree lengths with temperature
+    /// scaling, interpolating linearly between uniform and softmax distributions.
+    /// A temperature of 0.0 results in a uniform distribution, while a temperature of 1.0
+    /// results in a pure softmax distribution.
+    fn softmax(delta_tree_len: DVector<f64>, temperature: f64) -> DVector<f64> {
+        debug_assert!(
+            !delta_tree_len.is_empty(),
+            "The input vector must not be empty."
+        );
+
+        let mut exp_mat = Self::softmax_from_deltas(delta_tree_len);
+        let uniform_weight = 1.0 / exp_mat.len() as f64;
+
+        // Interpolated probabilities, temp = 0.0 is the uniform distribution, temp = 1.0 is the softmax of distances
+        // Avoid copying the matrix by mutating in place
+        for element in exp_mat.iter_mut() {
+            *element = ((1.0 - temperature) * uniform_weight) + ((temperature) * *element);
+        }
+        exp_mat
+    }
+
+    /// Computes a one-hot vector indicating the position(s) of the minimum value(s) in the input vector.
+    /// Used to uniformly select among the minimum delta tree lengths.
```

Comment:
Dont know if `one-hot` is an appropriate term here, since multiple indices can be 1. Usually with one-hot I expect a nested vector which with dimension [n_samples, n_classes].
Perhaps sth like this instead?
```suggestion
    /// Computes a vector indicating the position(s) of the minimum value(s) in the input vector,
    /// i.e., a vector with 1 at indices of the minimum, and 0 elsewhere.
    /// Used to uniformly select among the minimum delta tree lengths.
```

Reply:
Agreed, I think I went with the goddamn Copilot suggestion and didn't check the term's appropriateness here.